src/main/scala/gameengine/GameEngine.scala

```diff
@@ -0,0 +1,109 @@
+    // create a separator line
+    val separator = "|" + "-" * 17 + "+" + "-" * 12 + "+" + "-" * 17 + "+" + "-" * 17 + "|" + "|" + "-" * 17 + "+" + "-" * 12 + "+" + "-" * 17 + "+" + "-" * 17 + "|"
+    // create the values row
+    val row = f"| ${secretCharacter.name}%-15s | ${secretCharacter.gender}%-10s | ${secretCharacter.hairColor}%-15s | ${secretCharacter.wearsGlasses}%-15s |" +
+      f"${secretCharacter.wearsHat}%-15s${secretCharacter.hasBeard}%-15s${secretCharacter.eyeColor}%-15s"
+    val output: String = header + "\n" + separator + "\n" + row
+    output
+
+  }
+
+  /**
+   * Filters out the list of character based upon the question and answer
+   * @param characters The list of character for the gameBoard
+   * @param question   The answer to the question (true or false).
+   * @param answer     A list of characters that match the filter criteria.
+   * @return the filtered gameBoard for the player
+   */
+  def filterCharacters(characters: ListBuffer[Person], question: String, answer: Boolean): ListBuffer[Person] = {
+    question match {
+      case "Is your person male?" => characters.filterInPlace(_.gender == "Male" == answer)
+      case "Is your person female?" => characters.filterInPlace(_.gender == "Female" == answer)
+      case "Does your person have blonde hair?" => characters.filterInPlace(_.hairColor == "Blonde" == answer)
+      case "Does your person have brown hair?" => characters.filterInPlace(_.hairColor == "Brown" == answer)
+      case "Does your person have black hair?" => characters.filterInPlace(_.hairColor == "Black" == answer)
+      case "Does your person have red hair?" => characters.filterInPlace(_.hairColor == "Red" == answer)
+      case "Does your person have grey hair?" => characters.filterInPlace(_.hairColor == "Grey" == answer)
+      case "Does your person wear glasses?" => characters.filterInPlace(_.wearsGlasses == answer)
+      case "Does your person wear a hat?" => characters.filterInPlace(_.wearsHat == answer)
+      case "Does your person have a beard?" => characters.filterInPlace(_.hasBeard == answer)
+      case "Does your person have blue eyes?" => characters.filterInPlace(_.eyeColor == "Blue" == answer)
+      case "Does your person have green eyes?" => characters.filterInPlace(_.eyeColor == "Green" == answer)
+      case "Does your person have brown eyes?" => characters.filterInPlace(_.eyeColor == "Brown" == answer)
+      case "Does your person have hazel eyes?" => characters.filterInPlace(_.eyeColor == "Hazel" == answer)
+      case _ => characters.clear() // If the question does not match any case, clear the list
+    }
+    characters
+  }
+
+  /**
+   * Checks if the length of the gameBoard is 1 means only one character left in the gameBoard
+   * @param characters The game board of the player
+   * @return returns true if the length of gameBoard == 1 else returns false
+   */
+  def endGame(playerGameBoard: ListBuffer[Person]): Boolean = {
+    playerGameBoard.length == 1
```

Comment:
Sorry if I have just missed it but I cannot see anywhere you are using endGame.

Reply:
When thinking about how to end the game and keep the game going in the refactoring of guess who consider recursion.